eng/common/TestResources/New-TestResources.ps1

```diff
@@ -76,7 +76,14 @@ param (
     [switch] $Force,
 
     [Parameter()]
-    [switch] $OutFile
+    [switch] $OutFile,
+
+    [Parameter()]
+    [boolean] $OutFileEncryption = $true,
```

Comment:
This is what `[switch]` is for and is more idiomatic for PowerShell:

```suggestion
    [switch] $OutFileEncryption = $true,
```

Reply:
I'm not sure this should even be an option. From a security prospective it is never a good idea to output these to a unencrypted file.

Reply:
`$OutFile` already was encrypted on Windows. Come to think of it, this parameter seems redundant. You should use `$OutFile` as-is and refactor your code so that it works for other environments. The Windows method must remain as-is.

Reply:
I think file encryption as more of a defensive measure to prevent people from accidentally checking in secrets to git. I don't see how it adds much in the way of security, at least for Microsoft users. If you have access to the machine which can decrypt the secret, why does it matter whether it's encrypted or not? The stress cluster node filesystems are encrypted at rest, so this file is only plain text from the context of the container. Similarly developer hard drives are also encrypted by corporate device policy.

We may want a better way of preventing local users from inadvertently using this flag, but anyone really determined can just comment out a few lines in their powershell.

If we want to restrict AKS cluster users from being able to access secrets, then we can delegate azure resource creation to a separate namespaced service only accessible to those with specific cluster roles. The local stress test container would still have access to any credentials outputted by the ARM template, but ideally those would be resource group scoped. Albert is already going to work on resource deletion, so we could design some extra tasks to handle resource creation.

Perhaps another middle ground here is to add some support for linux file encryption, so linux users of the script (e.g. me) can generate an outfile for local runs?

Reply:
Responding to a [thread](https://github.com/Azure/azure-sdk-tools/pull/2053#discussion_r718927835) below but keeping it here to make this discussion easier to track.

- To specify ENV variables in kubernetes, they either show up as plain text in the container resource yaml or they are a reference to a kubernetes secret object, which you then need to create and maintain separately (which is more annoying to do dynamically from a pod vs. statically at creation time).
- In kubernetes/container environments, the volume mount in which these secrets are stored is tied to the lifecycle of the pod/container, which itself is tied to the lifecycle of the process ([emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume type). I have to do a little more research about how this behaves with pod `Completed` states when run within a job and whether it sticks around.
- All container files can be encrypted on the host: https://docs.microsoft.com/en-us/azure/aks/enable-host-encryption. I will go enable this for our clusters.

Reply:
Yes the encryption is definitely a defensive measure but it is an important one. There have definitely been a number of folks that accidently check-in files that have secrets in them and we don't want to add an mechanism that would make it easy for folks to accidently do that. 

As for trying to do file encryption on linux ourselves I would not suggest we try. The .NET team also will not try because it isn't really possible to get correct (https://github.com/dotnet/runtime/issues/22886) so it would be a losing battle for us to try. 

So the question comes down to how can we persist these values securely for usage in our Stress tests. The main goal is to avoid persisting these things to disk that could be leaked, is there a good way to temporarily capture the environment variable set by the script and set/pass them in the machines you are running? Do all the machine need the all of the secrets or only a subset that we could pass to them?

Reply:
For linux, @pakrym and I once discussed just doing what the Azure CLI does: chmod'd 700 dot file. If it's good enough for Azure CLI and other tools, there's no reason it's not good enough for this case. Don't store the file in the repo at all. In fact, I didn't want to store .NET's .env files there, but Pavel made a good point about otherwise having to invent some sort of semantic path lookup. Maybe that's what's needed here, though. Perhaps just hash the path to the file that would be in the repo, and store it under `~/.azuresdk/<hash>/.env` or something. Easily deterministic and repeatable so the test infrastructure can find it. Don't have to do that for every language. Let the `TestEnvironment` (e.g. in .NET) or whatever do the OS check and respond accordingly.

Reply:
And if we need to take a file path, we'll need a new parameter. Originally I argued that's what `-OutFile` is in every other cmdlet, but it got merged anyway so we'll have to find another one. Maybe `-OutFilePath`? That's assuming we need to even pass the path in, though. I wonder if we do given my previous comment, though. Do it consistently across platforms and it's still deterministic and "secure enough".

Reply:
The pointer indirection to another user scoped file might be an interesting idea. However one other measure the cli takes is that that have access tokens that expire, they don't have long lasting secrets. That is also another potential mitigation we can take but I suspect that may be difficult to monitor because all the resources have all sorts of different secret types.

Reply:
Just an idea, but we could have the New-TestResources script clean out old files based on last access date or something. Or, if we store the expiration in the tokens, base it on that. We can figure out some clean-up options here. Could even start it in a background job so it doesn't hold up the main process (though I suspect it would be really quick).

Reply:
I like the home directory+chmod idea, with a couple modifications:

1. I would write it to `$HOME/.local` as that plays a little nicer with [linux convention](https://www.freedesktop.org/software/systemd/man/file-hierarchy.html#Home%20Directory) and doesn't pollute the home directory with dot files/folders.
2. I think a hashed filepath is not going to be readable (let's say someone has a dozen of those from past runs, how do they know which one to source?). Perhaps using the resource group name the deployment is scoped to would be better, or even the full file path?

```
~/.local/azuresdk/rg-bebroder-storage.test-resources.env
# OR
~/.local/azuresdk/azure-sdk-for-net/sdk/appconfiguration/test-resources.env
```

Reply:
> So the question comes down to how can we persist these values securely for usage in our Stress tests. The main goal is to avoid persisting these things to disk that could be leaked, is there a good way to temporarily capture the environment variable set by the script and set/pass them in the machines you are running? Do all the machine need the all of the secrets or only a subset that we could pass to them?

One thing we can do is use [aad pod identity](https://github.com/Azure/aad-pod-identity) to bind either an MSI or service principal to the pod, and control test resource access via role bindings. The container process can grab a token for a resource via the instance metadata endpoint (see [example](https://github.com/Azure/aad-pod-identity/blob/master/examples/rest-api/client/script.sh#L10)).

This does not completely solve the problem because people will still need to output connection strings in their ARM template for the newly created test resources, but it would address persistence on the app container of the main provisioner identity which is most important. Regarding connection strings, I don't think there's a great way to avoid persisting these. Even env variables get "persisted" in that they are stored on the `/proc` filesystem, which will have the same duration as any volume mount since both are tied to the lifetime of the container. All that said, "persistence" in this context is still just "for as long as the process is active" since that's what the container lifetime is coupled to.

I don't see much of a risk of people checking in files that only exist on a container in `/mnt`, unless for some reason they are doing all their development in an ephemeral container hosted in the cluster, and including that directory in their git root. Whether these secrets are stored in a volume mount, env variable, or IMDS endpoint is the same from the perspective of an attacker (if you can SSH into the container you can read any of these, unless we remove a shell from the container images).

Logging is a bigger risk here and something we'll need to be more vigilant about. I think restricting the stress test app container to only having tokenized provisioner access (if it needs that at all) and connection strings for ephemeral resources is going to offer the lowest blast radius. I do worry about putting more burden on stress test writers, because the IMDS endpoint calls won't work for local development of the code. Perhaps our SDK libraries already handle this well?